### DISCUSSION TITLE
Are server actions guaranteed to run in succession?

### DISCUSSION THREAD

**Discussion by codinginflow** [vercel/next.js#51443] [App Router]
I have this code in my app that uses a server action to create a new (anonymous ) shopping cart if the visitor clicks an "add to cart" button and no cart for them exists yet:

    "use server";
    
    import ...
    
    export async function incrementProductQuantity(
      productId: string,
    ) {
      const session = await getServerSession(authOptions);
    
      const cart = (await getCart(session)) ?? (await createCart(session));
    
      ...add item...
    
      revalidatePath("/products/[id]");
    }

I was worried about **race conditions** when the button is clicked multiple times and it would create multiple new carts. But when I tried it out, the server action wasn't executed multiple times in parallel, but one after another.

So do server actions always execute in succession or is my thinking wrong here?

**Comment by bigoldcat123** [comment]
```
export const write_unsafe = async (_name:string) => {
    console.log('write_unsafe',_name);

    const r = await db.data_teacher.findFirst({
        where:{
            name:_name
        }
    })
    console.log(r);
    if(r) {return 'exist'} 
    console.log('write -> ',_name);
    
    await new Promise(resolve => setTimeout(resolve, 3000))
    const t = await db.data_teacher.create({
        data:{
            name:_name
        }
    })
    return t.name
}
```

---result---

write_unsafe hello
null
write ->  hello
write_unsafe hello
null
write ->  hello
 POST /dashboard/1996070003/650d4983-86e2-11ef-8e30-0242ac110002 200 in 3088ms
 POST /dashboard/1996070003/650d4983-86e2-11ef-8e30-0242ac110002 200 in 3031ms

I guess it is not successive😭
